Documentation/design-docs/hosting-layer-apis.md

```diff
@@ -0,0 +1,145 @@
+* `argc` / `argv` - command-line arguments
+* `buffer` - buffer to populate with the output (including a null terminator)
+* `buffer_size` - size of `buffer` in `char_t` units
+* `required_buffer_size` - if `buffer` is too small, this will be populated with the minimum required buffer size
+
+### .NET Core 3.0+
+
+``` C
+int corehost_get_coreclr_delegate(coreclr_delegate_type type, void **delegate)
+```
+
+Get a delegate for CoreCLR functionality
+* `type` - requested type of runtime functionality
+* `delegate` - function pointer to the requested runtime functionality
+
+``` C
+typedef void(*corehost_resolve_component_dependencies_result_fn)(
+    const char_t *assembly_paths,
+    const char_t *native_search_paths,
+    const char_t *resource_search_paths);
+
+int corehost_resolve_component_dependencies(
+    const char_t *component_main_assembly_path,
+    corehost_resolve_component_dependencies_result_fn result)
+```
+
+Resolve dependencies for the specified component.
+* `component_main_assembly_path` - path to the component
+* `result` - callback which will receive the results of the component dependency resolution
+
+See [Component dependency resolution support in host](host-component-dependencies-resolution.md)
+
+``` C
+typedef void(*corehost_error_writer_fn)(const char_t *message);
+
+corehost_error_writer_fn corehost_set_error_writer(corehost_error_writer_fn error_writer)
+```
+
+Set a callback which will be used to report error messages. By default no callback is registered and the errors are written to standard error.
+* `error_writer` - callback function which will be invoked every time an error is reported. When set to `nullptr`, this function unregisters any previously registered callback and the default behaviour is restored.
+
+The return value is the previouly registered callback (which is now unregistered) or `nullptr` if there was no previously registered callback.
+
+The error writer is registered per-thread. On each thread, only one callback can be registered. Subsequent registrations overwrite the previous ones.
+
+#### [Proposed] .NET Core 3.0+
+
+``` C
+typedef void* context_handle;
+
+struct corehost_context_contract
+{
+    size_t version;
+    context_handle instance;
+    int (*get_property)(
+        context_handle instance,
+        const char_t *key,
+        const char_t **value);
+    int (*set_property)(
+        context_handle instance,
+        const char_t *key,
+        const char_t *value);
+    int (*get_properties)(
+        context_handle instance,
+        size_t *count,
+        const char_t **keys,
+        const char_t **values);
+};
+```
+
+Contract for performing operations on an initialized host context.
+* `version` - version of the struct.
+* `instance` - opaque handle to the initialized host context.
```

Comment:
`initialized` once initialized isn't it too late to set properties?

Reply:
```suggestion
* `instance` - opaque handle to the `corehost_context_contract` state.
```

Reply:
The design is to run everything up to the point of loading `coreclr`, without actually loading it. So We can still modify properties.